### DISCUSSION TITLE
Can I pre-build images locally and just push them?

### DISCUSSION THREAD

**Discussion by nemanjam** [dokku/dokku#5530] [Q&A] [answered]
I have 1GB VPS that already has few containers plus Dokku container and naturally I noticed my server froze and was unable to build simple Node.js image. But I'm sure it would have enough resources to run 2-3 images more.

Is there a way to pre-build images locally and just push and deploy them to Dokku server?

**Comment by nemanjam** [comment]
I assume exactly that is described here, correct?

https://dokku.com/docs/deployment/methods/git/#initializing-an-app-repository-from-a-docker-image

But then Dokku comes down to simple router/reverse-proxy, right?

**Comment by josegonzalez** [comment, marked as answer]
You can use the `git:from-image` ([docs](https://dokku.com/docs/deployment/methods/git/#initializing-an-app-repository-from-a-docker-image)) command to deploy remote images. Those images can be built in any way you want - buildah, pack, nixpacks, whatever - and then Dokku will treat it as a normal image.

While this might seem like its devolving Dokku to managing just routing, Dokku also would manage:

- env vars
- datastores
- persistent storage
- cron
- scaling/resource management
- etc.

Even within routing, there are topics such as domain management and ssl.

That said, if you don't find that compelling, you are free to use anything else :) Since it's your server, I won't stop you from doing whatever you'd like.

---

Note: nodejs is pretty heavy to build. Heroku actually gives you much larger build servers than what it schedules them onto, which is why it seems things build quickly and regardless of your dyno size.

**Reply by nemanjam** [reply]
Thank you.
